File: src/Wording.Core/Learning/WordSelector.cs
namespace Wording.Core.Learning;

/// <summary>
/// Wybiera nastepne slowko do pokazania.
/// <para>
/// Celowo NIE bramkuje po terminie wymagalnosci, jak klasyczny SRS. Ta aplikacja
/// wyswietla slowko co kilka minut w tle, a nie w sesjach powtorek - przy sztywnych
/// terminach przez wiekszosc czasu nie mialaby czego pokazac. Zamiast tego losuje
/// z wagami: slowka przeterminowane dominuja, dobrze znane pojawiaja sie rzadko,
/// ale nigdy nie znikaja calkiem z rotacji.
/// </para>
/// </summary>
public static class WordSelector
{
    /// <summary>Slowko jeszcze nieocenione ma byc pokazywane czesto, zeby szybko weszlo do obiegu.</summary>
    internal const double NewWordWeight = 10.0;

    /// <summary>Waga slowka dokladnie w terminie. Kazdy dzien opoznienia dodaje 1.</summary>
    internal const double DueWeight = 1.0;

    /// <summary>Gorne ograniczenie opoznienia, zeby jedno zapomniane slowko sprzed roku nie zdominowalo rotacji.</summary>
    internal const double MaxOverdueDays = 30.0;

    /// <summary>Dolna waga slowka swiezo powtorzonego - mala, ale niezerowa, wiec nic nie wypada z rotacji.</summary>
    internal const double MinWeight = 0.02;

    /// <summary>Zwraca slowko do pokazania albo null, jesli lista jest pusta.</summary>
    public static Word? PickNext(IReadOnlyList<Word> words, DateTimeOffset now, Random random)
    {
        ArgumentNullException.ThrowIfNull(words);
        ArgumentNullException.ThrowIfNull(random);

        if (words.Count == 0)
        {
            return null;
        }

        var weights = new double[words.Count];
        var total = 0.0;

        for (var i = 0; i < words.Count; i++)
        {
            weights[i] = Weight(words[i], now);
            total += weights[i];
        }

        var roll = random.NextDouble() * total;

        for (var i = 0; i < weights.Length; i++)
        {
            roll -= weights[i];
            if (roll <= 0)
            {
                return words[i];
            }
        }

        // Nieosiagalne poza bledami zaokraglen na sumie zmiennoprzecinkowej.
        return words[^1];
    }

    /// <summary>Waga slowka: im pilniejsze, tym wieksza szansa na wylosowanie.</summary>
    internal static double Weight(Word word, DateTimeOffset now)
    {
        if (word.IsNew)
        {
            return NewWordWeight;
        }

        var overdueDays = (now - word.Review.DueUtc).TotalDays;

        if (overdueDays >= 0)
        {
            return DueWeight + Math.Min(overdueDays, MaxOverdueDays);
        }

        // Slowko jeszcze niewymagalne - waga maleje, im dalej do terminu.
        return Math.Max(MinWeight, DueWeight / (1 - overdueDays));
    }
}

File: macos/Sources/WordingKit/WordSelector.swift
import Foundation

/// Wybiera nastepne slowko do pokazania.
///
/// Port `Wording.Core.Learning.WordSelector`. Celowo NIE bramkuje po terminie
/// wymagalnosci, jak klasyczny SRS: aplikacja pokazuje slowko co kilka sekund
/// w tle, a nie w sesjach powtorek, wiec przy sztywnych terminach przez
/// wiekszosc czasu nie mialaby czego wyswietlic. Zamiast tego losuje z wagami.
public struct WordSelector: Sendable {
    /// Slowko jeszcze nigdy nieocenione ma byc pokazywane czesto.
    public static let newWordWeight = 10.0

    /// Waga slowka dokladnie w terminie. Kazdy dzien opoznienia dodaje 1.
    public static let dueWeight = 1.0

    /// Gorne ograniczenie opoznienia, zeby jedno zapomniane slowko sprzed roku
    /// nie zdominowalo rotacji.
    public static let maxOverdueDays = 30.0

    /// Dolna waga - mala, ale niezerowa, wiec nic nie wypada z rotacji.
    public static let minWeight = 0.02

    public init() {}

    /// Zwraca slowko do pokazania albo nil, jesli lista jest pusta.
    public func pickNext(
        from words: [Word],
        now: Date,
        using generator: inout some RandomNumberGenerator
    ) -> Word? {
        guard !words.isEmpty else { return nil }

        let weights = words.map { Self.weight(for: $0, now: now) }
        let total = weights.reduce(0, +)

        var roll = Double.random(in: 0..<total, using: &generator)

        for (index, weight) in weights.enumerated() {
            roll -= weight
            if roll <= 0 {
                return words[index]
            }
        }

        // Nieosiagalne poza bledami zaokraglen na sumie zmiennoprzecinkowej.
        return words.last
    }

    public func pickNext(from words: [Word], now: Date) -> Word? {
        var generator = SystemRandomNumberGenerator()
        return pickNext(from: words, now: now, using: &generator)
    }

    /// Waga slowka: im pilniejsze, tym wieksza szansa na wylosowanie.
    public static func weight(for word: Word, now: Date) -> Double {
        guard word.review.lastReviewedUtc != nil else {
            return newWordWeight
        }

        let overdueDays = now.timeIntervalSince(word.review.dueUtc)
            / SpacedRepetitionScheduler.secondsPerDay

        if overdueDays >= 0 {
            return dueWeight + min(overdueDays, maxOverdueDays)
        }

        // Slowko jeszcze niewymagalne - waga maleje, im dalej do terminu.
        return max(minWeight, dueWeight / (1 - overdueDays))
    }
}
